fluss-server/src/test/java/com/alibaba/fluss/server/replica/ReplicaManagerTest.java

```diff
@@ -1163,6 +1165,100 @@ void testStopReplica() throws Exception {
         replicaManager.getReplicaOrException(tb);
     }
 
+    @Test
+    void testKvDataVisibility() throws Exception {
+        // The CDC log is only visible after the KV has been flushed to RocksDB. In other words,
+        // when we can read the CDC log, the associated kv record must have been
+        // inserted/updated/deleted in RocksDB. The reason for ensuring this visibility is that we
+        // first buffer the data in memory before flushing it to RocksDB. Thus, we need to guarantee
+        // visibility.
+        TableBucket tb = new TableBucket(DATA1_TABLE_ID_PK, 1);
+        makeKvTableAsLeader(DATA1_TABLE_ID_PK, DATA1_TABLE_PATH_PK, tb.getBucket());
+        Replica replica = replicaManager.getReplicaOrException(tb);
+        RocksDBKv rocksDBKv = replica.getKvTablet().getRocksDBKv();
+        long beginTime = System.nanoTime();
+
+        CompactedKeyEncoder keyEncoder = new CompactedKeyEncoder(DATA1_ROW_TYPE, new int[] {0});
+        // retry send kv records to kv store, if the highWatermark increased, the kv record must be
+        // visible in rocksdb.
+        int round = 1000;
+        Thread writerThread =
+                new Thread(
+                        () -> {
+                            CompletableFuture<List<PutKvResultForBucket>> future;
+                            for (int i = 0; i < round; i++) {
+                                future = new CompletableFuture<>();
+                                Object[] key = {i};
+                                Object[] value = {i, "a"};
+                                // don't wait for the result.
+                                try {
+                                    replicaManager.putRecordsToKv(
+                                            20000,
+                                            -1,
+                                            Collections.singletonMap(
+                                                    tb,
+                                                    genKvRecordBatch(
+                                                            Collections.singletonList(
+                                                                    Tuple2.of(key, value)))),
+                                            null,
+                                            future::complete);
+                                } catch (Exception e) {
+                                    throw new RuntimeException(e);
+                                }
+                            }
+                        });
+
+        long[] highWatermarkUpdateTimestamps = new long[round];
+        Thread readWatermarkThead =
+                new Thread(
+                        () -> {
+                            int count = 0;
+                            while (count < round) {
+                                if (replica.getLogHighWatermark() == count + 1) {
```

Comment:
The test is unstable as the while loop may become a dead loop.

![image](https://github.com/user-attachments/assets/f9470fef-36d0-43df-9822-0a467622aff0)

Reply:
Done! Change `==` to `>=` to avoid dead loop